From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Sun, 18 Dec 2022 13:40:05 -0800
Subject: [PATCH] More DragonBattle API

== AT ==
public net.minecraft.world.level.dimension.end.EndDragonFight GATEWAY_COUNT
public net.minecraft.world.level.dimension.end.EndDragonFight gateways
public net.minecraft.world.level.dimension.end.EndDragonFight respawnCrystals
public net.minecraft.world.level.dimension.end.EndDragonFight spawnNewGateway(Lnet/minecraft/core/BlockPos;)V

diff --git a/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java b/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java
index f6f5fad52e41d5ccc37dc28c40904fce014907fb..6b59c4b5906c0fb4fdbc674452c6ff3df42b099b 100644
--- a/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -439,6 +439,24 @@ public class EndDragonFight {
         this.gateways.clear();
     }
 
+    // Paper start - More DragonBattle API
+    public boolean spawnNewGatewayIfPossible() {
+        if (!this.gateways.isEmpty()) {
+            this.spawnNewGateway();
+            return true;
+        }
+        return false;
+    }
+
+    public List<EndCrystal> getSpikeCrystals() {
+        final List<EndCrystal> endCrystals = new java.util.ArrayList<>();
+        for (final SpikeFeature.EndSpike spike : SpikeFeature.getSpikesForLevel(this.level)) {
+            endCrystals.addAll(this.level.getEntitiesOfClass(EndCrystal.class, spike.getTopBoundingBox()));
+        }
+        return endCrystals;
+    }
+    // Paper end - More DragonBattle API
+
     private void spawnNewGateway() {
         if (!this.gateways.isEmpty()) {
             int i = (Integer) this.gateways.remove(this.gateways.size() - 1);
diff --git a/src/main/java/org/bukkit/craftbukkit/boss/CraftDragonBattle.java b/src/main/java/org/bukkit/craftbukkit/boss/CraftDragonBattle.java
index dc7fdc120f4317ee1b7935ef226eddb0406aee6e..6bfabb38b51115beb2a65a165f235347838b6006 100644
--- a/src/main/java/org/bukkit/craftbukkit/boss/CraftDragonBattle.java
+++ b/src/main/java/org/bukkit/craftbukkit/boss/CraftDragonBattle.java
@@ -137,4 +137,46 @@ public class CraftDragonBattle implements DragonBattle {
     private DragonRespawnAnimation toNMSRespawnPhase(RespawnPhase phase) {
         return (phase != RespawnPhase.NONE) ? DragonRespawnAnimation.values()[phase.ordinal()] : null;
     }
+    // Paper start - More DragonBattle API
+    @Override
+    public int getGatewayCount() {
+        return EndDragonFight.GATEWAY_COUNT - this.handle.gateways.size();
+    }
+
+    @Override
+    public boolean spawnNewGateway() {
+        return this.handle.spawnNewGatewayIfPossible();
+    }
+
+    @Override
+    public void spawnNewGateway(final io.papermc.paper.math.Position position) {
+        this.handle.spawnNewGateway(io.papermc.paper.util.MCUtil.toBlockPos(position));
+    }
+
+    @Override
+    public java.util.List<org.bukkit.entity.EnderCrystal> getRespawnCrystals() {
+        if (this.handle.respawnCrystals == null) {
+            return java.util.Collections.emptyList();
+        }
+
+        final java.util.List<org.bukkit.entity.EnderCrystal> enderCrystals = new java.util.ArrayList<>();
+        for (final net.minecraft.world.entity.boss.enderdragon.EndCrystal endCrystal : this.handle.respawnCrystals) {
+            if (!endCrystal.isRemoved() && endCrystal.isAlive() && endCrystal.valid) {
+                enderCrystals.add(((org.bukkit.entity.EnderCrystal) endCrystal.getBukkitEntity()));
+            }
+        }
+        return java.util.Collections.unmodifiableList(enderCrystals);
+    }
+
+    @Override
+    public java.util.List<org.bukkit.entity.EnderCrystal> getHealingCrystals() {
+        final java.util.List<org.bukkit.entity.EnderCrystal> enderCrystals = new java.util.ArrayList<>();
+        for (final net.minecraft.world.entity.boss.enderdragon.EndCrystal endCrystal : this.handle.getSpikeCrystals()) {
+            if (!endCrystal.isRemoved() && endCrystal.isAlive() && endCrystal.valid) {
+                enderCrystals.add(((org.bukkit.entity.EnderCrystal) endCrystal.getBukkitEntity()));
+            }
+        }
+        return java.util.Collections.unmodifiableList(enderCrystals);
+    }
+    // Paper end - More DragonBattle API
 }
